### DISCUSSION TITLE
Precedence problems with a "catch all" rule

### DISCUSSION THREAD

**Discussion by dimus** [tree-sitter/tree-sitter#1407] [Q&A] [answered]
I have a grammar like this:

```javascript
terminator = '\n';

module.exports = grammar({
  name: 'sciname',

  rules: {
    source_file: ($) => repeat($._sci_name),

    _sci_name: ($) => seq($.uninomial, optional($.tail), terminator),

    _uninomial: ($) => /[A-Z][a-z]+/,

    uninomial: ($) => choice($._uninomial, seq($.genus, $.subgenus)),

    genus: ($) => $._uninomial,

    subgenus: ($) => seq('(', $._uninomial, ')'),

    tail: ($) => /.+/,
  },
});
```

`tail` rule suppose to catch anything that comes after `uninomial`. However I have trouble with this test:

```txt
=====================
Subgenus
=====================

Bubo (Bubo)

---------------------

(source_file (uninomial (genus) (subgenus)))
```

producing: 

`(source_file (uninomial) (tail))` 

instead of 

`(source_file (uninomial (genus) (subgenus)))`

Changing precedence does not seem to influence this behavior. Is there a way to make a "catch-all" rule like `tail` to engage only if all other previous rules failed?

**Comment by ahlinc** [comment, marked as answer]
Try to assign a lower lexer precedence to your `tail` with a `token(prec(-1, ...))` construct like:
```js
tail: ($) => token(prec(-1, /[^\s]+/)),
```
Related excerpt from [conflicting-tokens](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens) documentation:
> 2. Lexical Precedence - **When the precedence functions described above are used within the token function, the given precedence values serve as instructions to the lexer.** If there are two valid tokens that match the characters at a given position in the document, Tree-sitter will select the one with the higher precedence.

What means `token(prec(N, ...))` construction.

**Reply by dimus** [reply]
worked perfeclty, thank you @ahlinc, it did not occur to me that a precedence can be negative
